File: src/main/java/fi/quanfoxes/parser/nodes/ReturnNode.java
package fi.quanfoxes.parser.nodes;

import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Contextable;
import fi.quanfoxes.parser.Function;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Resolvable;
import fi.quanfoxes.parser.Resolver;
import fi.quanfoxes.parser.Type;

public class ReturnNode extends Node implements Resolvable {
    public ReturnNode(Node object) {
        super.add(object);
    }

    private Type getReturnType(Node node) throws Exception {
        if (node instanceof Contextable) {
            Contextable contextable = (Contextable)node;
            return (Type)contextable.getContext();
        }

        throw new Exception("Couldn't resolve return type");
    }

    @Override
    public Node resolve(Context context) throws Exception {
        // Returned object must be resolved first
        Node node = first();

        if (node instanceof Resolvable) {
            Resolvable resolvable = (Resolvable)node;
            Node resolved = resolvable.resolve(context);

            node.replace(resolved);
            node = resolved;
        }
        
        // Find the parent function where the return value can be assigned
        Function function = context.getFunctionParent();

        Type current = function.getReturnType();
        Type type = getReturnType(node);

        if (type == null) {
            throw new Exception("Couldn't resolve return type");
        }
        
        if (current != type) {
            Type shared = type;

            if (current != null) {
                shared = Resolver.getSharedType(current, type);
            }

            if (shared == null) {
                throw new Exception(String.format("Type '%s' isn't compatible with the current return type '%s'", type.getName(), current.getName()));
            }
    
            function.setReturnType(type);
        }
        
        return null;
    }
}

File: Zigzag/Parser/Nodes/ReturnNode.cs
using System;

public class ReturnNode : InstructionNode, IResolvable
{
	private Status CurrentStatus = Status.Error("Return type was not resolved");

	public Node Value => First!;

	public ReturnNode(Node node) : base(Keywords.RETURN)
	{
		Add(node);
	}

	private static Type? GetReturnType(Node node)
	{
		if (node is IType type)
		{
			return type.GetType();
		}

		return Types.UNKNOWN;
	}

	public Node? Resolve(Context context)
	{
		// Returned object must be resolved first
		var node = First;

		if (node == null)
		{
			CurrentStatus = Status.OK;
			return null;
		}

		Resolver.Resolve(context, node);

		// Find the parent function where the return value can be assigned
		var function = context.GetFunctionParent() ?? throw new ApplicationException("Return statement was not inside a function");

		var current = function.ReturnType;
		var type = GetReturnType(node ?? throw new ApplicationException("Return statment didn't have a value to return"));

		if (type == Types.UNKNOWN)
		{
			CurrentStatus = Status.Error("Could not resolve return type");
			return null;
		}

		if (current != type)
		{
			var shared = (Type?)type;

			if (current != Types.UNKNOWN)
			{
				shared = Resolver.GetSharedType(current, type);
			}

			if (shared == null)
			{
				CurrentStatus = Status.Error($"Type '{type.Name}' isn't compatible with the current return type '{current?.Name ?? "none"}'");
				return null;
			}

			function.ReturnType = type;
			CurrentStatus = Status.OK;
		}

		CurrentStatus = Status.OK;
		return null;
	}

	public override NodeType GetNodeType()
	{
		return NodeType.RETURN_NODE;
	}

	public Status GetStatus()
	{
		return CurrentStatus;
	}
}